text/0000-custom-dst.md

```diff
@@ -0,0 +1,479 @@
+    fn align_of_val((): ()) -> usize { align_of::<T>() }
+}
+```
+
+If you have a type which you would like to be unsized,
+you can implement this trait for your type!
+
+```rust
+#[repr(C)]
+struct CStr([c_char; 0]);
+
+unsafe impl DynamicallySized for CStr {
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { strlen(&self.0 as *const c_char) + 1 }
+    fn align_of_val((): ()) -> usize { 1 }
+}
+```
+
+and your type will be `!Sized`.
+
+The existing `DynamicallySized` types will continue to work;
+if one writes a `DynamicallySized` type `T`,
+and then wraps `T` into a struct, they'll get the obvious semantics.
+
+```rust
+struct Foo {
+    x: usize,
+    y: CStr,
+}
+
+// size_of_val(&foo) returns size_of_header::<Foo>() + size_of_val(&foo.y)
+// same with align_of_val - simply `align_of_header::<Foo>()`
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+By language trait, we mean that `DynamicallySized` is a lang item.
+
+In addition to the explanation given above,
+we will also introduce three functions into the standard library,
+in `core::raw`, which allow you to create and destructure these
+pointers to `DynamicallySized` types:
+
+```rust
+mod core::raw {
+    pub fn from_raw_parts<T: ?Sized>(
+        ptr: *const (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *const T;
+
+    pub fn from_raw_parts_mut<T: ?Sized>(
+        ptr: *mut (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *mut T;
+
+    pub fn metadata<T: ?Sized>(
+        ptr: *const T,
+    ) -> <T as DynamicallySized>::Metadata;
+}
+```
+
+and we will introduce two functions into `core::mem`,
+to help people write types with Flexible Array Members:
+
+```rust
+mod core::mem {
+    pub fn size_of_header<T: ?DynamicallySized>() -> usize;
+    pub fn align_of_header<T: ?DynamicallySized>() -> usize;
+}
+```
+
+These functions return the size and alignment of the header of a type;
+or, the minimum possible size and alignment, in other words.
+For existing `Sized` types, they are equivalent to `size_of` and `align_of`,
+and for existing DSTs,
+
+```rust
+assert_eq!(size_of_header::<[T]>(), 0);
+assert_eq!(align_of_header::<[T]>(), align_of::<T>());
+assert_eq!(size_of_header::<dyn Trait>(), 0);
+assert_eq!(align_of_header::<dyn Trait>(), 1);
+
+// on 64-bit
+struct RcBox<T: ?Sized> {
+  strong: Cell<usize>,
+  weak: Cell<usize>,
+  value: T,
+}
+assert_eq!(size_of_header::<RcBox<dyn Trait>>(), 16);
+assert_eq!(align_of_header::<RcBox<dyn Trait>>(), 8);
+```
+
+Note that this is a minimum - this means that for `extern type`s,
+they return `0` and `1` respectively.
+
+Notes:
+  - names of the above functions should be bikeshed
+  - `extern type`s do not implement `DynamicallySized`, although in theory one
+    could choose to implement the trait for them
+    (that usecase is not supported by this RFC).
+  - `DynamicallySized` is a new trait in the `Sized` hierarchy
```

Comment:
What are the tradeoffs between this and requiring that every type implement `DynamicallySized`? Are there examples aside from `extern type` that shouldn't implement `DynamicallySized`?